searchcore/src/vespa/searchcore/proton/server/bucketmovejobv2.cpp

```diff
@@ -0,0 +1,368 @@
+    _diskMemUsageNotifier.addDiskMemUsageListener(this);
+    recompute();
+}
+
+BucketMoveJobV2::~BucketMoveJobV2()
+{
+    _bucketCreateNotifier.removeListener(this);
+    _clusterStateChangedNotifier.removeClusterStateChangedHandler(this);
+    _bucketStateChangedNotifier.removeBucketStateChangedHandler(this);
+    _diskMemUsageNotifier.removeDiskMemUsageListener(this);
+}
+
+BucketMoveJobV2::NeedResult
+BucketMoveJobV2::needMove(const ScanIterator &itr) const {
+    NeedResult noMove(false, false);
+    const bool hasReadyDocs = itr.hasReadyBucketDocs();
+    const bool hasNotReadyDocs = itr.hasNotReadyBucketDocs();
+    if (!hasReadyDocs && !hasNotReadyDocs) {
+        return noMove; // No documents for bucket in ready or notready subdbs
+    }
+    const bool isActive = itr.isActive();
+    // No point in moving buckets when node is retired and everything will be deleted soon.
+    // However, allow moving of explicitly activated buckets, as this implies a lack of other good replicas.
+    if (!_calc || (_calc->nodeRetired() && !isActive)) {
+        return noMove;
+    }
+    const bool shouldBeReady = _calc->shouldBeReady(document::Bucket(_bucketSpace, itr.getBucket()));
+    const bool wantReady = shouldBeReady || isActive;
+    LOG(spam, "checkBucket(): bucket(%s), shouldBeReady(%s), active(%s)",
+        itr.getBucket().toString().c_str(), bool2str(shouldBeReady), bool2str(isActive));
+    if (wantReady) {
+        if (!hasNotReadyDocs)
+            return noMove; // No notready bucket to make ready
+    } else {
+        if (isActive)
+            return noMove; // Do not move rom ready to not ready when active
+        if (!hasReadyDocs)
+            return noMove; // No ready bucket to make notready
+    }
+    return {true, wantReady};
+}
+
+void
+BucketMoveJobV2::startMove(BucketMoverSP mover, size_t maxDocsToMove) {
+    auto [keys, done] = mover->getKeysToMove(maxDocsToMove);
+    if (done) {
+        mover->setBucketDone();
+    }
+    if (keys.empty()) return;
+    if (_stopped.load(std::memory_order_relaxed)) return;
+    mover->updateLastValidGid(keys.back()._gid);
+    auto context = getLimiter().beginOperation();
+    Bucket spiBucket(document::Bucket(_bucketSpace, mover->getBucket()));
+    auto bucketTask = makeBucketTask(
+            [this, mover=std::move(mover), keys=std::move(keys),opsTracker=getLimiter().beginOperation()]
+            (const Bucket & bucket, std::shared_ptr<IDestructorCallback> onDone) mutable
+    {
+        assert(mover->getBucket() == bucket.getBucketId());
+        using DoneContext = vespalib::KeepAlive<std::pair<IDestructorCallbackSP, IDestructorCallbackSP>>;
+        prepareMove(std::move(mover), std::move(keys),
+                    std::make_shared<DoneContext>(std::make_pair(std::move(opsTracker), std::move(onDone))));
+    });
+    auto failed = _bucketExecutor.execute(spiBucket, std::move(bucketTask));
+    if (!failed) {
+        _startedCount.fetch_add(1, std::memory_order_relaxed);
```

Comment:
This might leave _`executedCount` higher than `_startedCount` for a short time, but the anomaly should be harmless as long as inSync is only called from the master thread.

Reply:
Yes, and it will eventually converge.